File: day30/day_thirty.cpp
# include <iostream>
# include <queue>

using namespace std;

typedef struct Node
{
	int data;
	struct Node* left;
	struct Node* right;
}node;

int itr = 0;
int arr[] = {7,5,10,3,6,9,12};
int size = sizeof(arr)/sizeof(arr[0]);


node* getNode()
{
	node* newnode = (node*)malloc(sizeof(node));
	return newnode;
}

void readNode(node* newnode)
{
	newnode->data = arr[itr];
	itr = itr + 1;
	newnode->left = NULL;
	newnode->right = NULL;
}

node* insertNode(node* btree,node* newnode)
{
	if(btree == NULL)
	{
		return newnode;
	}
	else if(newnode->data < btree->data)
	{
		btree->left = insertNode(btree->left,newnode);
	}
	else if(newnode->data > btree->data)
	{
		btree->right = insertNode(btree->right,newnode);
	}
	else if(btree->data == newnode->data)
	{
		cout << "data already exist" << endl;
		return btree;
	}
	return btree;
}

node* createBtree()
{
	node* btree = NULL;
	for(int i=0;i<size;i++)
	{
		node* newnode = getNode();
		readNode(newnode);
		btree = insertNode(btree,newnode);
	}
	return btree;
}

int distanceFromRoot(node* newnode,int x)
{
	if(x < newnode->data)
	{
		return 1 + distanceFromRoot(newnode->left,x);
	}
	else if(x > newnode->data)
	{
		return 1 + distanceFromRoot(newnode->right,x);
	}

	return 0;
}

int distanceBetween2Node(node* newnode,int x,int y)
{
	if(newnode == NULL)
	{
		return 0;
	}

	if(x > y)
	{
		swap(x,y);
	}

	if(x < newnode->data && y < newnode->data)
	{
		return distanceBetween2Node(newnode->left,x,y);
	}

	if(x > newnode->data && y > newnode->data)
	{
		return distanceBetween2Node(newnode->right,x,y);
	}

	if(x <= newnode->data && y >= newnode->data)
	{
		return distanceFromRoot(newnode,x) + distanceFromRoot(newnode,y);
	}

	return 0;
}

int main()
{
	node* btree = createBtree();
	int x,y;

	x = 3;y = 9;
	cout << "distanceBetween2Node(" << x << "," << y << ") => " << distanceBetween2Node(btree,x,y) << endl;
	x = 3;y = 6;
	cout << "distanceBetween2Node(" << x << "," << y << ") => " << distanceBetween2Node(btree,x,y) << endl;

	return 0;
}

File: day30/DayThirty.java
import java.util.Queue;
import java.util.LinkedList;

public class DayThirty
{
	class Node
	{
		int data;
		Node left;
		Node right;
		private Node(int data)
		{
			this.data = data;
			this.left = null;
			this.right = null;
		}
	}

	private int itr = 0;
	private int arr[] = {7,5,10,3,6,9,12};

	private Node btree = null;
	
	private Node createBtree()
	{
		for(int itr=0;itr<arr.length;itr++)
		{
			Node newnode = new Node(arr[itr]);
			btree = insertNode(btree,newnode);
		}
		return btree;
	}

	private Node insertNode(Node btree,Node newnode)
	{
		if(btree == null)
		{
			return newnode;
		}
		else if(newnode.data < btree.data)
		{
			btree.left = insertNode(btree.left,newnode);
		}
		else if(newnode.data > btree.data)
		{
			btree.right = insertNode(btree.right,newnode);
		}
		else if(newnode.data == btree.data)
		{
			System.out.print("Data Already Present");
			return btree;
		}
		return btree;
	}

	private int distanceFromRoot(Node newnode,int x)
	{
		if(x > newnode.data)
		{
			return 1 + distanceFromRoot(newnode.right,x);
		}
		else if(x < newnode.data)
		{
			return 1 + distanceFromRoot(newnode.left,x);
		}

		return 0;
	}

	private int distanceFrom2Nodes(Node root,int x,int y)
	{
		if(root == null)
		{
			return 0;
		}

		if(x > y)
		{
			int temp = x;
			x = y;
			y = temp;
		}

		if(x > root.data && y > root.data)
		{
			return distanceFrom2Nodes(root.right,x,y);
		}

		if(x < root.data && y < root.data)
		{
			return distanceFrom2Nodes(root.left,x,y);
		}

		if(x <= root.data && y >= root.data)
		{
			return distanceFromRoot(root,x) + distanceFromRoot(root,y);
		}

		return 0;
	}

	public static void main(String args[])
	{
		DayThirty tree = new DayThirty();
		Node btree = tree.createBtree();

		int x,y;

		x = 3;y = 9;
		System.out.println("distanceBetween2Node(" + x + "," + y + ") => " + tree.distanceFrom2Nodes(btree,x,y));
		x = 3;y = 6;
		System.out.println("distanceBetween2Node(" + x + "," + y + ") => " + tree.distanceFrom2Nodes(btree,x,y));
	}
}